### PR TITLE
Add BitCheck Process

### PR DESCRIPTION
**Pull request by mgkwill**
Issue Number: #801

Objective of pull request: Add a BitCheck process. This short and simple ProcessModel can be used for quick checking of bit-accurate process runs as to whether bits will overflow when running on hardware.

## Pull request checklist

Your PR fulfills the following requirements:
-   [X] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
-   [X] Tests are part of the PR (for bug fixes / features)
-   [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
-   [X] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
-   [X] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
-   [X] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
-   [X] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:
-   [X] Feature

## What is the current behavior?

- Currently there is not a standard way to check if process vars overflow particular bit limitations during bit-accurate runs meant to simulate hardware.

## What is the new behavior?

- BitCheck process is added that gives a standard way to check if process vars overflow particular bit limitations during bit-accurate runs meant to simulate hardware.

## Does this introduce a breaking change?

-   [ ] Yes
-   [X] No

## Supplemental information
During Loihi verification against Lava bit-accurate simulation we found that it was useful to have an overflow checker. This PR and feature request is the result of that realization.

Suggestions and design ideas happily accepted.

### CODE REVIEW

**Review by bamsumit — APPROVED** [review]
Looks great Marcus. Feel free to incorporate the dynamic shape identification during `connect_var` that we discussed.

**Comment by mgkwill** [comment]
> Looks great Marcus. Feel free to incorporate the dynamic shape identification during `connect_var` that we discussed.

Thanks @bamsumit will make that change.

**Comment by mgkwill** [comment]
> Looks great Marcus. Feel free to incorporate the dynamic shape identification during `connect_var` that we discussed.

Hi @bamsumit - I've refactored to set shape using connect_var. Let me know if what I've done works, or you have any other thoughts?

**Comment by bamsumit** [comment]
> > Looks great Marcus. Feel free to incorporate the dynamic shape identification during `connect_var` that we discussed.
> 
> Hi @bamsumit - I've refactored to set shape using connect_var. Let me know if what I've done works, or you have any other thoughts?

@mgkwill That is perfect :)

File: src/lava/proc/bit_check/models.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                print(
+                    "min signed value {}".format(
+                        self.max_signed_int_per_bits(self.bits)
+                    )
+                )
+
+    def check_bit_overflow(self, value: ty.Type[np.ndarray]):
+        value = value.astype(np.int32)
+        shift_amt = 32 - self.bits
+        # shift value left by shift_amt and
```
Comment (PhilippPlank) [inline src/lava/proc/bit_check/models.py:86]: Comments should start with upper case letter.

File: src/lava/proc/bit_check/process.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    ) -> None:
+        """BitCheck process.
+        This process is used for quick checking of
+        bit-accurate process run as to whether bits will
+        overflow when running on bit sensitive hardware.
+
+        Parameters
+        ----------
+        shape: Tuple
+            shape of the sigma process.
```
Comment (PhilippPlank) [inline src/lava/proc/bit_check/process.py:31]: Start sentence with uppercase letter.

**Review by PhilippPlank — APPROVED** [review]

**Comment by mgkwill** [comment]
Thanks @PhilippPlank and @bamsumit for the feedback!

**Pull request merged**